crates/history/src/utils.rs

```diff
@@ -1,11 +1,24 @@
 use std::cell::RefCell;
 use std::rc::{Rc, Weak};
 use std::sync::atomic::{AtomicU32, Ordering};
+use std::sync::Once;
 
 use wasm_bindgen::throw_str;
 
 pub(crate) fn get_id() -> u32 {
     static ID_CTR: AtomicU32 = AtomicU32::new(0);
+    static INIT: Once = Once::new();
+
+    INIT.call_once(|| {
+        // Use the most significant 32 bits of `now`'s mantissa and reverse it.
+        // The time 1696926809097.0 is stored as something like 0x1.8b18b83c09000 * 2^0x28,
+        // and the start value for it is 0x3c1d18d1 (0x8b18b83c reversed).
+        // This way the numbers change significantly when a little time is passed.
+        // It's not a perfect solutions but should reduce collisions significantly.
+        let start = ((js_sys::Date::now().to_bits() >> (52 - 32)) as u32).reverse_bits();
```

Comment:
We should not use JavaScript APIs to seed the counter because `MemoryHistory` can be used in SSR (non-WebAssembly targets) as well.

I think it is better to use [getrandom](https://docs.rs/getrandom/latest/getrandom/fn.getrandom.html#) as we don't have to guarantee that an Id is `u32` or sequential.

Reply:
I can split it and let MemoryHistory use the old (and for it correct behavior) and only use this for BrowserHistory.
Or use gertandom for both or just BrowserHistory.

Reply:
I think we can use `getrandom` for `wasm32` and existing behaviour for other platforms as I feel it is safer to deal with collision using a cryptographically secure entropy. Having `MemoryHistory` to use sequential slots whilst other histories use random slots might increase the chance of collision. We compare the id of 2 locations to determine of whether 2 locations are the same. So it is better to also protect against collision across history types.

`getrandom` uses `window.crypto.getRandomValues` on WebAssembly, which should have minimal bundle size increase for Web targets and performance is not a concern (as that is usually limited by how many clicks a user can do).

The current behaviour should be significantly faster than `getrandom` (and `rand::thread_rng().fill`) for SSR, and cannot be restored by browser so we can keep the existing behaviour for other platforms.

We can make location id to be an opaque type such as: `struct LocationId([u8; 8])` as `u64` is outside of JavaScript's safe integer range.
I think u64 or larger is better to prevent collision.

Reply:
As a first step I created a new version which uses getrandom on wasm and leaves it unchanged on all other platforms.

As for the collision: getrandom is still only used to pick the starting number and then increase by one, which removes collision within the lifespan of the app, which is a good idea.
Which is why I thought to depend on reversed time, that way it changes a lot when little time has passed.

Should I now change the result of get_id from `u32` into `struct LocationId([u8; 8])`, and thus the inner counter from u32 ti u64? Also in the case of BrowserHistory it's used more as an `StateId` and in a `HashMap<u32, Rc<dyn Any>>` and the `LocationId/StateId` is stored in javascript via `serde_wasm_bindgen::to_value`. While MemoryHistory uses it only for tracking a location within rust.

Reply:
I think it is fine to keep it as u32 for now, so we can publish a patch release to fix the issue and then make this breaking change in the next minor release.